src/lib.rs

```diff
@@ -181,8 +179,13 @@ pub struct Machine<'tsunami> {
     /// equivalent to `public_ip`.
     pub public_dns: String,
 
+    #[cfg(feature = "openssh")]
```

Comment:
This won't work since cargo features have to be additive. You'll want to add `#[non_exhausive]` to the struct, then we should be fine.

Reply:
Could you explain why this is non-additive? I've also read a bit about `#[non_exhaustive]` but couldn't understand how it plays with cargo features.

Reply:
Ah, actually, in this particular case the private `_tsunami` field saves us, because it essentially emulates `#[non_exhausive]`. But I still think it'd be good to add. The basic idea is that a user _could_ have written this code (if all the fields were public):
```rust
let Machine {
  nickname,
  public_ip,
  private_ip,
  public_dns,
  username,
  private_key,
} = vm;
```
In a crate, call it `A`, that depends on `tsunami` _without_ the `openssh` feature. Now imagine some other crate, `B`, which also depends on `tsunami`, but _with_ the `openssh` feature. In isolation, they both compile just fine. But now imagine some crate `C` that depends on both `A` and `B`. Cargo only builds one instances of every crate (per major version), so `tsunami` will only be built once. And since `B` requires the `openssh` feature, `tsunami` will be built with that feature enabled (this is the additive part — Cargo takes the _union_ of all the feature requirements). But this will make `A` not compile — the compiler will complain (correctly) that the pattern match above is missing the field `ssh`. The same problem applies to crates that construct a `Machine` by explicitly listing all its fields.

`#[non_exhausive]` makes it so that a type cannot be constructed using its implicit constructor (i.e., by listing its fields), and so that it cannot be matched against without including the wildcard pattern (`..` for fields, `_` for enum variants).

Reply:
That makes total sense now. Thanks for the awesome reply!

Reply:
I'm glad! The other thing we have to be careful with here is that the added fields cannot change any aspect of the containing type. For example, it would not be okay if including `openssh::Session` made the type `!Sync` or `!Send` (I don't believe that's the case here), because otherwise you end up in non-additive land again. Specifically, imagine that `A` depends on `Machine` being `Send`, but it stops being `Send` once `B` adds the `openssh` feature. `A` would then stop compiling.